edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs

```diff
@@ -0,0 +1,631 @@
+                {
+                    await this.SendResponseAsync(HttpStatusCode.Unauthorized);
+                    return;
+                }
+
+                IEdgeHub edgeHub = await this.edgeHubGetter;
+                IDeviceScopeIdentitiesCache identitiesCache = edgeHub.GetDeviceScopeIdentitiesCache();
+                Option<string> targetAuthChain = await identitiesCache.GetAuthChain(targetDeviceId);
+                if (!targetAuthChain.HasValue)
+                {
+                    Events.AuthorizationFail_NoAuthChain(requestData.Module.DeviceId);
+                    await this.SendResponseAsync(HttpStatusCode.Unauthorized);
+                    return;
+                }
+
+                string edgeDeviceId = edgeHub.GetEdgeDeviceId();
+                RegistryApiHttpResult result = await this.apiClient.PutModuleAsync(
+                    edgeDeviceId,
+                    new CreateOrUpdateModuleOnBehalfOfData($"{targetAuthChain.OrDefault()}", requestData.Module),
+                    ifMatchHeader);
+                await this.SendResponseAsync(result.StatusCode, result.JsonContent);
+                Events.CompleteRequest(nameof(this.CreateOrUpdateModuleOnBehalfOfAsync), edgeDeviceId, targetAuthChain.OrDefault(), result);
+            }
+            catch (Exception ex)
+            {
+                Events.InternalServerError(nameof(this.CreateOrUpdateModuleOnBehalfOfAsync), ex);
+                await this.SendResponseAsync(HttpStatusCode.InternalServerError, FormatErrorResponseMessage(ex.ToString()));
+            }
+        }
+
+        [HttpPost]
+        [Route("devices/{actorDeviceId}/modules/$edgeHub/getModuleOnBehalfOf")]
+        public async Task GetModuleOnBehalfOfAsync(
+            [FromRoute] string actorDeviceId,
+            [FromBody] GetModuleOnBehalfOfData requestData)
+        {
+            try
+            {
+                Events.ReceivedOnBehalfOfRequest(nameof(this.GetModuleOnBehalfOfAsync), actorDeviceId, Events.GetAdditionalInfo(requestData));
+
+                try
+                {
+                    Preconditions.CheckNonWhiteSpace(actorDeviceId, nameof(actorDeviceId));
+                    Preconditions.CheckNotNull(requestData, nameof(requestData));
+                    Preconditions.CheckNonWhiteSpace(requestData.AuthChain, nameof(requestData.AuthChain));
+                    Preconditions.CheckNonWhiteSpace(requestData.ModuleId, nameof(requestData.ModuleId));
+                }
+                catch (Exception ex)
+                {
+                    Events.BadRequest(nameof(this.GetModuleOnBehalfOfAsync), ex.Message);
+                    await this.SendResponseAsync(HttpStatusCode.BadRequest, FormatErrorResponseMessage(ex.ToString()));
+                    return;
+                }
+
+                actorDeviceId = WebUtility.UrlDecode(actorDeviceId);
+                if (!AuthChainHelpers.TryGetTargetDeviceId(requestData.AuthChain, out string targetDeviceId))
```

Comment:
>TryGetTargetDeviceId [](start = 38, length = 20)

same as above, this should always be an auth-chain for a device as opposed to module.

Reply:
same reply as above.